zetta_utils/mazepa_layer_processing/segmentation/adjust_affinities_with_masks.py

```diff
@@ -0,0 +1,109 @@
+from typing import Iterable
+
+import torch
+
+from zetta_utils import builder
+
+
+@builder.register("pass_all_data_if_any_masked")
+def pass_all_data_if_any_masked(
+    src: torch.Tensor,
+    masks: Iterable[torch.Tensor],
+) -> torch.Tensor:
+    """
+    If any mask is present, return the data, otherwise return zeros
+
+    :param src: input Tensor
+    :param masks: list of masks
+    """
+    is_masked = any(mask.sum() > 0 for mask in masks)
+    has_data = src.sum() > 0
+    result = src if is_masked and has_data else torch.zeros_like(src)
+    return result
+
+
+@builder.register("adjust_affinities_with_masks")
+def adjust_affinities_with_masks(
+    src: torch.Tensor,
+    semantic_mask: torch.Tensor,
+    threshold_mask: torch.Tensor,
+    threshold_value: float = 0.85,
+    fill_value: float = 0,
+) -> torch.Tensor:
+    """
+    Post-processing for affinities before segmentation
+
+    :param src: input Tensor
+    :param semantic_mask: mask that will adjust affinities across mask boundaries
+    :param threshold_mask: mask that will adjust affinities below `threshold_value`
+    :param threshold_value: value for thresholding
+    :param fill_value: value to set the adjusted affinities
+    """
+    aff_thresholded = torch.zeros_like(src)
+    if (semantic_mask.sum() > 0) or (threshold_mask.sum() > 0):
+        aff_mask_boundary = adjust_affinities_across_mask_boundary(
+            src=src, mask=semantic_mask, fill_value=fill_value
+        )
+        aff_thresholded = adjust_thresholded_affinities_in_mask(
+            src=aff_mask_boundary,
+            mask=threshold_mask,
+            threshold_value=threshold_value,
+            fill_value=fill_value,
+        )
+    return aff_thresholded
+
+
+@builder.register("adjust_affinities_across_mask_boundary")
+def adjust_affinities_across_mask_boundary(
+    src: torch.Tensor,
+    mask: torch.Tensor,
+    fill_value: float = 0,
+) -> torch.Tensor:
+    result = src
+    """
+    Adjust affinities that span masked and unmasked voxel pairs.
+
+    Note: At each voxel, we store the affinities to neighboring voxels in
+    the negative cardinal directions. Meaning that location (1, 1, 1) in
+    the affinity map contains the following affinities:
+    (0, 1, 1) - (1, 1, 1)
+    (1, 0, 1) - (1, 1, 1)
+    (1, 1, 0) - (1, 1, 1)
+
+    :param src: input Tensor
+    :param mask: mask matching dimensions of src
+    :param fill_value: value to set the adjusted affinities
+    """
+    mask_compare = mask[0, 1:, :, :] != mask[0, :-1, :, :]
+    aff_mask = torch.nn.functional.pad(mask_compare, (0, 0, 0, 0, 1, 0))
+    result[0, aff_mask] = fill_value
+    mask_compare = mask[0, :, 1:, :] != mask[0, :, :-1, :]
+    aff_mask = torch.nn.functional.pad(mask_compare, (0, 0, 1, 0, 0, 0))
+    result[1, aff_mask] = fill_value
+    mask_compare = mask[0, :, :, 1:] != mask[0, :, :, :-1]
+    aff_mask = torch.nn.functional.pad(mask_compare, (1, 0, 0, 0, 0, 0))
```

Comment:
See comment below, but it might be better to factor this part out into `masks.py`